presto-main/src/main/java/com/facebook/presto/operator/HttpRpcShuffleClient.java

```diff
@@ -65,17 +66,25 @@
 
     private final HttpClient httpClient;
     private final URI location;
+    private final Optional<URI> asyncPageTransportLocation;
 
     public HttpRpcShuffleClient(HttpClient httpClient, URI location)
+    {
+        this(httpClient, location, Optional.empty());
+    }
+
+    public HttpRpcShuffleClient(HttpClient httpClient, URI location, Optional<URI> asyncPageTransportLocation)
     {
         this.httpClient = requireNonNull(httpClient, "httpClient is null");
         this.location = requireNonNull(location, "location is null");
+        this.asyncPageTransportLocation = requireNonNull(asyncPageTransportLocation, "asyncPageTransportLocation is null");
     }
 
     @Override
     public ListenableFuture<PagesResponse> getResults(long token, DataSize maxResponseSize)
     {
-        URI uri = uriBuilderFrom(location).appendPath(String.valueOf(token)).build();
+        URI uriBase = asyncPageTransportLocation.orElse(location);
+        URI uri = uriBuilderFrom(uriBase).appendPath(String.valueOf(token)).build();
```

Comment:
Why don't we leave this class basically unchanged and just do this `ExchangeClient` (conditionally change the location based on the config property)?

Reply:
`location` is unchanged for other request type like `acknowledgeResult`

Reply:
Well, we could modify it just before we create this client.  We're already switching off the URI in `ExchangeClient`, so some light URI parsing/rewriting wouldn't be totally out of place.

Reply:
We didn't switch off URI in `ExchangeClient`, we conditionally add another location based on config property.

Reply:
Perhaps we're not understanding each other, all I'm suggesting is this:

```java
            case "http":
            case "https":
                resultClient = new HttpRpcShuffleClient(httpClient, asyncPageTransportLocation.orElse(location));
                break;
```

There's no reason to do this inside the `HttpRpcShuffleClient`.